File: android-app/app/src/main/java/com/medicscientist/calculator/calculator/Parser.kt
package com.medicscientist.calculator.calculator

/**
 * Parser - Recursive descent parser that builds an AST
 * Equivalent to parser.c in the C version
 */

sealed class ASTNode {
    data class NumberNode(val value: Double) : ASTNode()
    data class ConstantNode(val value: Double) : ASTNode()
    
    data class BinaryOpNode(
        val op: BinaryOp,
        val left: ASTNode,
        val right: ASTNode
    ) : ASTNode()
    
    data class UnaryOpNode(
        val op: UnaryOp,
        val operand: ASTNode
    ) : ASTNode()
    
    data class FunctionNode(
        val func: FunctionType,
        val argument: ASTNode
    ) : ASTNode()
}

enum class BinaryOp { ADD, SUBTRACT, MULTIPLY, DIVIDE, POWER }
enum class UnaryOp { NEGATE, FACTORIAL }
enum class FunctionType { SIN, COS, TAN, ASIN, ACOS, ATAN, LOG, LN, SQRT, ABS }

class Parser(private val tokens: List<Token>) {
    private var position = 0
    var hasError = false
    var errorMessage = ""
    
    private val currentToken: Token
        get() = if (position < tokens.size) tokens[position] else Token(TokenType.EOF)
    
    private fun advance() {
        if (position < tokens.size) position++
    }
    
    private fun expect(type: TokenType): Boolean {
        if (currentToken.type == type) {
            advance()
            return true
        }
        hasError = true
        errorMessage = "Expected $type but got ${currentToken.type}"
        return false
    }
    
    private fun parsePrimary(): ASTNode? {
        if (hasError) return null
        
        return when (currentToken.type) {
            TokenType.NUMBER -> {
                val node = ASTNode.NumberNode(currentToken.value)
                advance()
                node
            }
            TokenType.PI, TokenType.E -> {
                val node = ASTNode.ConstantNode(currentToken.value)
                advance()
                node
            }
            TokenType.LPAREN -> {
                advance()
                val node = parseExpression()
                expect(TokenType.RPAREN)
                node
            }
            TokenType.MINUS -> {
                advance()
                val operand = parsePrimary() ?: return null
                ASTNode.UnaryOpNode(UnaryOp.NEGATE, operand)
            }
            TokenType.PLUS -> {
                advance()
                parsePrimary()
            }
            TokenType.SIN, TokenType.COS, TokenType.TAN,
            TokenType.ASIN, TokenType.ACOS, TokenType.ATAN,
            TokenType.LOG, TokenType.LN, TokenType.SQRT, TokenType.ABS -> {
                val funcType = when (currentToken.type) {
                    TokenType.SIN -> FunctionType.SIN
                    TokenType.COS -> FunctionType.COS
                    TokenType.TAN -> FunctionType.TAN
                    TokenType.ASIN -> FunctionType.ASIN
                    TokenType.ACOS -> FunctionType.ACOS
                    TokenType.ATAN -> FunctionType.ATAN
                    TokenType.LOG -> FunctionType.LOG
                    TokenType.LN -> FunctionType.LN
                    TokenType.SQRT -> FunctionType.SQRT
                    TokenType.ABS -> FunctionType.ABS
                    else -> FunctionType.SIN
                }
                advance()
                if (!expect(TokenType.LPAREN)) return null
                val argument = parseExpression() ?: return null
                if (!expect(TokenType.RPAREN)) return null
                ASTNode.FunctionNode(funcType, argument)
            }
            else -> {
                hasError = true
                errorMessage = "Unexpected token: ${currentToken.type} ('${currentToken.lexeme}')"
                null
            }
        }?.let { node ->
            // Check for postfix factorial
            if (currentToken.type == TokenType.FACTORIAL) {
                advance()
                ASTNode.UnaryOpNode(UnaryOp.FACTORIAL, node)
            } else {
                node
            }
        }
    }
    
    private fun parseExponent(): ASTNode? {
        val left = parsePrimary() ?: return null
        if (currentToken.type == TokenType.POWER) {
            advance()
            val right = parseExponent() ?: return null // Right-associative
            return ASTNode.BinaryOpNode(BinaryOp.POWER, left, right)
        }
        return left
    }
    
    private fun parseTerm(): ASTNode? {
        var left = parseExponent() ?: return null
        while (currentToken.type == TokenType.MULTIPLY || currentToken.type == TokenType.DIVIDE) {
            val op = if (currentToken.type == TokenType.MULTIPLY) BinaryOp.MULTIPLY else BinaryOp.DIVIDE
            advance()
            val right = parseExponent() ?: return null
            left = ASTNode.BinaryOpNode(op, left, right)
        }
        return left
    }
    
    fun parseExpression(): ASTNode? {
        var left = parseTerm() ?: return null
        while (currentToken.type == TokenType.PLUS || currentToken.type == TokenType.MINUS) {
            val op = if (currentToken.type == TokenType.PLUS) BinaryOp.ADD else BinaryOp.SUBTRACT
            advance()
            val right = parseTerm() ?: return null
            left = ASTNode.BinaryOpNode(op, left, right)
        }
        return left
    }
    
    fun parse(): ASTNode? {
        val ast = parseExpression()
        if (currentToken.type != TokenType.EOF && !hasError) {
            hasError = true
            errorMessage = "Unexpected token after expression: '${currentToken.lexeme}'"
        }
        return ast
    }
}

File: c-source/parser.c
#include "parser.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Initialize parser with input string
void parser_init(Parser* parser, const char* input) {
    tokenizer_init(&parser->tokenizer, input);
    parser->current_token = tokenizer_next(&parser->tokenizer);
    parser->has_error = false;
    parser->error_message[0] = '\0';
}

// Set parser error
static void parser_error(Parser* parser, const char* message) {
    if (!parser->has_error) {
        parser->has_error = true;
        strncpy(parser->error_message, message, 255);
        parser->error_message[255] = '\0';
    }
}

// Advance to next token
static void advance(Parser* parser) {
    parser->current_token = tokenizer_next(&parser->tokenizer);
}

// Expect a specific token type
static bool expect(Parser* parser, TokenType type) {
    if (parser->current_token.type == type) {
        advance(parser);
        return true;
    }
    char msg[256];
    snprintf(msg, sizeof(msg), "Expected %s but got %s", 
             token_type_to_string(type), token_type_to_string(parser->current_token.type));
    parser_error(parser, msg);
    return false;
}

// Create a new AST node
static ASTNode* create_node(NodeType type) {
    ASTNode* node = (ASTNode*)malloc(sizeof(ASTNode));
    if (node == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        exit(1);
    }
    node->type = type;
    return node;
}

// Forward declarations for recursive descent
static ASTNode* parse_primary(Parser* parser);
static ASTNode* parse_exponent(Parser* parser);
static ASTNode* parse_term(Parser* parser);
static ASTNode* parse_additive(Parser* parser);

// Parse primary expressions (numbers, constants, functions, parenthesized expressions)
static ASTNode* parse_primary(Parser* parser) {
    if (parser->has_error) return NULL;
    
    ASTNode* node = NULL;
    
    switch (parser->current_token.type) {
        case TOKEN_NUMBER: {
            node = create_node(NODE_NUMBER);
            node->data.number_value = parser->current_token.value;
            advance(parser);
            break;
        }
        
        case TOKEN_PI: {
            node = create_node(NODE_CONSTANT);
            node->data.constant_value = parser->current_token.value;
            advance(parser);
            break;
        }
        
        case TOKEN_E: {
            node = create_node(NODE_CONSTANT);
            node->data.constant_value = parser->current_token.value;
            advance(parser);
            break;
        }
        
        case TOKEN_LPAREN: {
            advance(parser); // consume '('
            node = parse_expression(parser);
            if (!expect(parser, TOKEN_RPAREN)) {
                free_ast(node);
                return NULL;
            }
            break;
        }
        
        case TOKEN_MINUS: {
            // Unary minus
            advance(parser);
            ASTNode* operand = parse_primary(parser);
            if (operand == NULL) return NULL;
            node = create_node(NODE_UNARY_OP);
            node->data.unary.op = OP_NEGATE;
            node->data.unary.operand = operand;
            break;
        }
        
        case TOKEN_PLUS: {
            // Unary plus (just skip it)
            advance(parser);
            node = parse_primary(parser);
            break;
        }
        
        // Functions
        case TOKEN_SIN:
        case TOKEN_COS:
        case TOKEN_TAN:
        case TOKEN_ASIN:
        case TOKEN_ACOS:
        case TOKEN_ATAN:
        case TOKEN_LOG:
        case TOKEN_LN:
        case TOKEN_SQRT:
        case TOKEN_ABS: {
            FunctionType func;
            switch (parser->current_token.type) {
                case TOKEN_SIN:  func = FUNC_SIN;  break;
                case TOKEN_COS:  func = FUNC_COS;  break;
                case TOKEN_TAN:  func = FUNC_TAN;  break;
                case TOKEN_ASIN: func = FUNC_ASIN; break;
                case TOKEN_ACOS: func = FUNC_ACOS; break;
                case TOKEN_ATAN: func = FUNC_ATAN; break;
                case TOKEN_LOG:  func = FUNC_LOG;  break;
                case TOKEN_LN:   func = FUNC_LN;   break;
                case TOKEN_SQRT: func = FUNC_SQRT; break;
                case TOKEN_ABS:  func = FUNC_ABS;  break;
                default:         func = FUNC_SIN;  break;
            }
            advance(parser); // consume function name
            
            if (!expect(parser, TOKEN_LPAREN)) {
                return NULL;
            }
            
            ASTNode* argument = parse_expression(parser);
            if (argument == NULL) return NULL;
            
            if (!expect(parser, TOKEN_RPAREN)) {
                free_ast(argument);
                return NULL;
            }
            
            node = create_node(NODE_FUNCTION);
            node->data.function.func = func;
            node->data.function.argument = argument;
            break;
        }
        
        default: {
            char msg[256];
            snprintf(msg, sizeof(msg), "Unexpected token: %s ('%s')", 
                     token_type_to_string(parser->current_token.type),
                     parser->current_token.lexeme);
            parser_error(parser, msg);
            return NULL;
        }
    }
    
    // Check for postfix factorial
    if (node != NULL && parser->current_token.type == TOKEN_FACTORIAL) {
        advance(parser);
        ASTNode* factorial_node = create_node(NODE_UNARY_OP);
        factorial_node->data.unary.op = OP_FACTORIAL;
        factorial_node->data.unary.operand = node;
        node = factorial_node;
    }
    
    return node;
}

// Parse exponentiation (right-associative)
static ASTNode* parse_exponent(Parser* parser) {
    ASTNode* left = parse_primary(parser);
    if (left == NULL || parser->has_error) return left;
    
    if (parser->current_token.type == TOKEN_POWER) {
        advance(parser);
        ASTNode* right = parse_exponent(parser); // Right-associative recursion
        if (right == NULL) {
            free_ast(left);
            return NULL;
        }
        
        ASTNode* node = create_node(NODE_BINARY_OP);
        node->data.binary.op = OP_POWER;
        node->data.binary.left = left;
        node->data.binary.right = right;
        return node;
    }
    
    return left;
}

// Parse multiplication and division
static ASTNode* parse_term(Parser* parser) {
    ASTNode* left = parse_exponent(parser);
    if (left == NULL || parser->has_error) return left;
    
    while (parser->current_token.type == TOKEN_MULTIPLY || 
           parser->current_token.type == TOKEN_DIVIDE) {
        BinaryOp op;
        if (parser->current_token.type == TOKEN_MULTIPLY) {
            op = OP_MULTIPLY;
        } else {
            op = OP_DIVIDE;
        }
        advance(parser);
        
        ASTNode* right = parse_exponent(parser);
        if (right == NULL) {
            free_ast(left);
            return NULL;
        }
        
        ASTNode* node = create_node(NODE_BINARY_OP);
        node->data.binary.op = op;
        node->data.binary.left = left;
        node->data.binary.right = right;
        left = node;
    }
    
    return left;
}

// Parse addition and subtraction
static ASTNode* parse_additive(Parser* parser) {
    ASTNode* left = parse_term(parser);
    if (left == NULL || parser->has_error) return left;
    
    while (parser->current_token.type == TOKEN_PLUS || 
           parser->current_token.type == TOKEN_MINUS) {
        BinaryOp op;
        if (parser->current_token.type == TOKEN_PLUS) {
            op = OP_ADD;
        } else {
            op = OP_SUBTRACT;
        }
        advance(parser);
        
        ASTNode* right = parse_term(parser);
        if (right == NULL) {
            free_ast(left);
            return NULL;
        }
        
        ASTNode* node = create_node(NODE_BINARY_OP);
        node->data.binary.op = op;
        node->data.binary.left = left;
        node->data.binary.right = right;
        left = node;
    }
    
    return left;
}

// Main entry point for parsing
ASTNode* parse_expression(Parser* parser) {
    return parse_additive(parser);
}

// Free AST memory
void free_ast(ASTNode* node) {
    if (node == NULL) return;
    
    switch (node->type) {
        case NODE_NUMBER:
        case NODE_CONSTANT:
            break;
        case NODE_BINARY_OP:
            free_ast(node->data.binary.left);
            free_ast(node->data.binary.right);
            break;
        case NODE_UNARY_OP:
            free_ast(node->data.unary.operand);
            break;
        case NODE_FUNCTION:
            free_ast(node->data.function.argument);
            break;
    }
    
    free(node);
}

// Print AST for debugging
void print_ast(ASTNode* node, int indent) {
    if (node == NULL) return;
    
    for (int i = 0; i < indent; i++) printf("  ");
    
    switch (node->type) {
        case NODE_NUMBER:
            printf("Number: %.6f\n", node->data.number_value);
            break;
        case NODE_CONSTANT:
            printf("Constant: %.6f\n", node->data.constant_value);
            break;
        case NODE_BINARY_OP: {
            const char* ops[] = {"+", "-", "*", "/", "^"};
            printf("BinaryOp: %s\n", ops[node->data.binary.op]);
            print_ast(node->data.binary.left, indent + 1);
            print_ast(node->data.binary.right, indent + 1);
            break;
        }
        case NODE_UNARY_OP: {
            if (node->data.unary.op == OP_NEGATE) {
                printf("UnaryOp: -\n");
            } else {
                printf("UnaryOp: !\n");
            }
            print_ast(node->data.unary.operand, indent + 1);
            break;
        }
        case NODE_FUNCTION: {
            const char* funcs[] = {"sin", "cos", "tan", "asin", "acos", "atan", "log", "ln", "sqrt", "abs"};
            printf("Function: %s\n", funcs[node->data.function.func]);
            print_ast(node->data.function.argument, indent + 1);
            break;
        }
    }
}
